GopherCAN.c

```diff
@@ -402,92 +332,167 @@ S8 send_can_command(PRIORITY priority, MODULE_ID dest_module, GCAN_COMMAND_ID co
-//  function to directly send a CAN message with the specified parameter to
-//  another module
+// encodes and sends the specified parameter's group in a standard 11-bit CAN frame
 // params:
-//  PRIORITY priority:        PRIO_LOW or PRIO_HIGH
-//  MODULE_ID dest_module:    what module to send the parameter to
-//  GCAN_PARAM_ID parameter:  what parameter to send
+// CAN_INFO_STRUCT* param: parameter to send (along with its group)
 // returns:
-//  error codes specified in GopherCAN.h
-S8 send_parameter(PRIORITY priority, MODULE_ID dest_module, GCAN_PARAM_ID parameter)
+// error codes specified in GopherCAN.h
+S8 send_parameter(CAN_INFO_STRUCT* param)
 {
-	CAN_ID id;
-	CAN_MSG message;
-	U64 data = 0;
-	S8 c;
-	FLOAT_CONVERTER float_con;
-
-	// make sure the parameter is valid
-	if (parameter <= CAN_COMMAND_ID || parameter >= NUM_OF_PARAMETERS)
-	{
-		return BAD_PARAMETER_ID;
-	}
-
-	// make sure the parameter is enabled
-	if (!((CAN_INFO_STRUCT*)(all_parameter_structs[parameter]))->update_enabled)
-	{
-		return NOT_ENABLED_ERR;
-	}
-
-	// build the return message ID
-	id.priority = priority;
-	id.dest_module = dest_module;
-	id.source_module = this_module_id;
-	id.error = FALSE;
-	id.parameter = parameter;
-
-	build_message_id(&message, &id);
-
-	// set the RTR bit to data type
-	message.rtr_bit = DATA_MESSAGE;
-
-	// get the value of the data on this module and build the CAN message
-	if (parameter_data_types[parameter] == UNSIGNED8
-		|| parameter_data_types[parameter] == SIGNED8)
-	{
-		data |= ((U8_CAN_STRUCT*)(all_parameter_structs[parameter]))->data;
-		message.dlc = sizeof(U8);
-	}
-
-	else if (parameter_data_types[parameter] == UNSIGNED16
-		|| parameter_data_types[parameter] == SIGNED16)
-	{
-		data |= ((U16_CAN_STRUCT*)(all_parameter_structs[parameter]))->data;
-		message.dlc = sizeof(U16);
-	}
-
-	else if (parameter_data_types[parameter] == UNSIGNED32
-		|| parameter_data_types[parameter] == SIGNED32)
-	{
-		data |= ((U32_CAN_STRUCT*)(all_parameter_structs[parameter]))->data;
-		message.dlc = sizeof(U32);
-	}
+    PARAM_GROUP* group = NULL;
+
+    // find the specified parameter group
+    for (U8 i = 0; i < NUM_OF_GROUPS; i++) {
+        if (GROUPS[i].id == param->GROUP_ID) {
+            group = &GROUPS[i];
+            break;
+        }
+    }
+
+    if (group == NULL) return NOT_FOUND_ERR;
+
+    // build parameter group message
+    CAN_MSG message;
+
+    message.header.StdId = param->GROUP_ID;
+    message.header.IDE = CAN_ID_STD;
+    message.header.RTR = DATA_MESSAGE;
+    message.header.DLC = 8;
+
+    // encode parameters
+    U8 param_start = 0;
+    U8 param_length = 1;
+    S8 err;
+    for (U8 i = 1; i < 8; i++) {
+        if (group->slots[i] == group->slots[i-1]) {
+            // count consecutive slots belonging to the same parameter
+            param_length++;
+        } else {
+            // start of a new parameter, encode previous and write to data
+            GCAN_PARAM_ID id = group->slots[i-1];
```

Comment:
How does this code handle groups that are not filled with all 8 bytes?

Reply:
EMPTY_ID is just 0 if it's between two parameters. If the end of a data field has EMPTY_IDs, they get trimmed off like before.

Reply:
Cool